pkg/kubelet/container/helpers_test.go

```diff
@@ -306,3 +307,58 @@ func TestMakePortMappings(t *testing.T) {
 		assert.Equal(t, tt.expectedPortMappings, actual, "[%d]", i)
 	}
 }
+
+func TestHashContainer(t *testing.T) {
+	tests := []struct {
+		name         string
+		container1   *v1.Container
+		container2   *v1.Container
+		expectedHash uint64
+	}{
+		{
+			name: "testing redundant empty/nil field value",
+			container1: &v1.Container{
+				Name:  "test_container",
+				Image: "foo/image:v1",
+				Args: []string{
+					"/bin/sh",
+					"-c",
+					"echo abc",
+				},
+				Ports: []v1.ContainerPort{{ContainerPort: 8001}},
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceName(v1.ResourceStorage): resource.MustParse("10Gi"),
+					},
+				},
+			},
+			container2: &v1.Container{
+				Name:  "test_container",
+				Image: "foo/image:v1",
+				Args: []string{
+					"/bin/sh",
+					"-c",
+					"echo abc",
+				},
+				Ports: []v1.ContainerPort{{ContainerPort: 8001}},
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceName(v1.ResourceStorage): resource.MustParse("10Gi"),
+					},
+				},
+				// adding empty string and nil pointer
+				WorkingDir: "",
+				Lifecycle:  nil,
+			},
+			expectedHash: uint64(0xaac6921d),
+		},
+	}
+
+	for _, tt := range tests {
+		hashVal1 := HashContainer(tt.container1)
+		assert.Equal(t, tt.expectedHash, hashVal1, "%s: mismatched hash value", tt.name)
+
+		hashVal2 := HashContainer(tt.container2)
```

Comment:
I believe the container1 and container2 have exactly the same content and they will have the same hash without or without your fix. I don't think the test is working as you intended...

Reply:
@yujuhong Right.

Seems it's hard to reproduce the issue in the testcase.

Reply:
yeah, you need two different versions of a go struct, which you can't do in a single build

Reply:
the way I'd probably test this is to take your new HashContainer method, and compute a hash for a stubbed v1.Container in 1.15, then compare that to the same stubbed container in 1.16. The prior implementation would differ because new fields were added to the struct. The new implementation should remain the same since the json serializations would not have changed